File: backend/src/main/java/org/luxons/sevenwonders/game/boards/Board.java
package org.luxons.sevenwonders.game.boards;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.cards.Card;
import org.luxons.sevenwonders.game.cards.Color;
import org.luxons.sevenwonders.game.effects.SpecialAbility;
import org.luxons.sevenwonders.game.resources.Production;
import org.luxons.sevenwonders.game.resources.TradingRules;
import org.luxons.sevenwonders.game.scoring.PlayerScore;
import org.luxons.sevenwonders.game.scoring.ScoreCategory;
import org.luxons.sevenwonders.game.wonders.Wonder;

public class Board {

    private final Wonder wonder;

    private final int playerIndex;

    private final List<Card> playedCards = new ArrayList<>();

    private final Production production = new Production();

    private final Production publicProduction = new Production();

    private final Science science = new Science();

    private final TradingRules tradingRules;

    private final Military military;

    private final Set<SpecialAbility> specialAbilities = EnumSet.noneOf(SpecialAbility.class);

    private Map<Integer, Boolean> consumedFreeCards = new HashMap<>();

    private Card copiedGuild;

    private int gold;

    private int pointsPer3Gold;

    public Board(Wonder wonder, int playerIndex, Settings settings) {
        this.wonder = wonder;
        this.playerIndex = playerIndex;
        this.gold = settings.getInitialGold();
        this.tradingRules = new TradingRules(settings.getDefaultTradingCost());
        this.military = new Military(settings.getLostPointsPerDefeat(), settings.getWonPointsPerVictoryPerAge());
        this.pointsPer3Gold = settings.getPointsPer3Gold();
        this.production.addFixedResource(wonder.getInitialResource(), 1);
        this.publicProduction.addFixedResource(wonder.getInitialResource(), 1);
    }

    public Wonder getWonder() {
        return wonder;
    }

    public List<Card> getPlayedCards() {
        return playedCards;
    }

    public void addCard(Card card) {
        playedCards.add(card);
    }

    int getNbCardsOfColor(List<Color> colorFilter) {
        return (int) playedCards.stream().filter(c -> colorFilter.contains(c.getColor())).count();
    }

    public boolean isPlayed(String cardName) {
        return getPlayedCards().stream().map(Card::getName).filter(name -> name.equals(cardName)).count() > 0;
    }

    public Production getProduction() {
        return production;
    }

    public Production getPublicProduction() {
        return publicProduction;
    }

    public TradingRules getTradingRules() {
        return tradingRules;
    }

    public Science getScience() {
        return science;
    }

    public int getGold() {
        return gold;
    }

    public void setGold(int amount) {
        this.gold = amount;
    }

    public void addGold(int amount) {
        this.gold += amount;
    }

    public void removeGold(int amount) {
        if (gold < amount) {
            throw new InsufficientFundsException(gold, amount);
        }
        this.gold -= amount;
    }

    public Military getMilitary() {
        return military;
    }

    public void addSpecial(SpecialAbility specialAbility) {
        specialAbilities.add(specialAbility);
    }

    public boolean hasSpecial(SpecialAbility specialAbility) {
        return specialAbilities.contains(specialAbility);
    }

    public boolean canPlayFreeCard(int age) {
        return hasSpecial(SpecialAbility.ONE_FREE_PER_AGE) && !consumedFreeCards.getOrDefault(age, false);
    }

    public void consumeFreeCard(int age) {
        consumedFreeCards.put(age, true);
    }

    public void setCopiedGuild(Card copiedGuild) {
        if (copiedGuild.getColor() != Color.PURPLE) {
            throw new IllegalArgumentException("The given card '" + copiedGuild + "' is not a Guild card");
        }
        this.copiedGuild = copiedGuild;
    }

    public Card getCopiedGuild() {
        return copiedGuild;
    }

    public PlayerScore computePoints(Table table) {
        PlayerScore score = new PlayerScore(gold);
        score.put(ScoreCategory.CIVIL, computePointsForCards(table, Color.BLUE));
        score.put(ScoreCategory.MILITARY, military.getTotalPoints());
        score.put(ScoreCategory.SCIENCE, science.computePoints());
        score.put(ScoreCategory.TRADE, computePointsForCards(table, Color.YELLOW));
        score.put(ScoreCategory.GUILD, computePointsForCards(table, Color.PURPLE));
        score.put(ScoreCategory.WONDER, wonder.computePoints(table, playerIndex));
        score.put(ScoreCategory.GOLD, computeGoldPoints());
        return score;
    }

    private int computePointsForCards(Table table, Color color) {
        return playedCards.stream()
                          .filter(c -> c.getColor() == color)
                          .flatMap(c -> c.getEffects().stream())
                          .mapToInt(e -> e.computePoints(table, playerIndex))
                          .sum();
    }

    private int computeGoldPoints() {
        return gold / 3 * pointsPer3Gold;
    }

    static class InsufficientFundsException extends RuntimeException {
        InsufficientFundsException(int current, int required) {
            super(String.format("Current balance is %d gold, but %d are required", current, required));
        }
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/boards/Board.kt
package org.luxons.sevenwonders.game.boards

import org.luxons.sevenwonders.game.Player
import org.luxons.sevenwonders.game.Settings
import org.luxons.sevenwonders.game.cards.Card
import org.luxons.sevenwonders.game.cards.Color
import org.luxons.sevenwonders.game.data.Age
import org.luxons.sevenwonders.game.effects.SpecialAbility
import org.luxons.sevenwonders.game.resources.Production
import org.luxons.sevenwonders.game.resources.TradingRules
import org.luxons.sevenwonders.game.score.PlayerScore
import org.luxons.sevenwonders.game.score.ScoreCategory
import org.luxons.sevenwonders.game.wonders.Wonder

class Board internal constructor(val wonder: Wonder, val playerIndex: Int, settings: Settings) {

    val production = Production()
    val publicProduction = Production()
    val science = Science()
    val tradingRules: TradingRules = TradingRules(settings.defaultTradingCost)
    val military: Military = Military(settings.lostPointsPerDefeat, settings.wonPointsPerVictoryPerAge)
    private val pointsPer3Gold: Int = settings.pointsPer3Gold

    private val playedCards: MutableList<Card> = arrayListOf()
    private val specialAbilities: MutableSet<SpecialAbility> = hashSetOf()
    private val consumedFreeCards: MutableMap<Age, Boolean> = mutableMapOf()

    var gold: Int = settings.initialGold
        private set

    var copiedGuild: Card? = null
        internal set(copiedGuild) {
            if (copiedGuild!!.color !== Color.PURPLE) {
                throw IllegalArgumentException("The given card '$copiedGuild' is not a Guild card")
            }
            field = copiedGuild
        }

    init {
        this.production.addFixedResource(wonder.initialResource, 1)
        this.publicProduction.addFixedResource(wonder.initialResource, 1)
    }

    fun getPlayedCards(): List<Card> = playedCards

    internal fun addCard(card: Card) {
        playedCards.add(card)
    }

    internal fun getNbCardsOfColor(colorFilter: List<Color>): Int = playedCards.count { colorFilter.contains(it.color) }

    fun isPlayed(cardName: String): Boolean = playedCards.count { it.name == cardName } > 0

    internal fun addGold(amount: Int) {
        this.gold += amount
    }

    internal fun removeGold(amount: Int) {
        if (gold < amount) {
            throw InsufficientFundsException(gold, amount)
        }
        this.gold -= amount
    }

    internal fun addSpecial(specialAbility: SpecialAbility) {
        specialAbilities.add(specialAbility)
    }

    fun hasSpecial(specialAbility: SpecialAbility): Boolean = specialAbilities.contains(specialAbility)

    fun canPlayFreeCard(age: Age): Boolean =
        hasSpecial(SpecialAbility.ONE_FREE_PER_AGE) && !consumedFreeCards.getOrDefault(age, false)

    fun consumeFreeCard(age: Age) {
        consumedFreeCards[age] = true
    }

    internal fun computeScore(player: Player): PlayerScore = PlayerScore(
        boardGold = gold,
        pointsByCategory = mapOf(
            ScoreCategory.CIVIL to computePointsForCards(player, Color.BLUE),
            ScoreCategory.MILITARY to military.totalPoints,
            ScoreCategory.SCIENCE to science.computePoints(),
            ScoreCategory.TRADE to computePointsForCards(player, Color.YELLOW),
            ScoreCategory.GUILD to computePointsForCards(player, Color.PURPLE),
            ScoreCategory.WONDER to wonder.computePoints(player),
            ScoreCategory.GOLD to computeGoldPoints()
        )
    )

    private fun computePointsForCards(player: Player, color: Color): Int =
        playedCards.filter { it.color === color }
            .flatMap { it.effects }
            .map { it.computePoints(player) }
            .sum()

    private fun computeGoldPoints(): Int = gold / 3 * pointsPer3Gold

    internal class InsufficientFundsException(current: Int, required: Int) :
        IllegalStateException("Current balance is $current gold, but $required are required")
}
